### PR TITLE
Added tests for F-24E fixed placement scenarios

### PR DESCRIPTION
**Pull request by aditipande96**
Signed-off-by: Aditi Pande <aditi.pande@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: libs/csm/rest/csm_rest_capacity.py [hunk 1]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
...
+            total=total_written)
+        elif failure_cnt == kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=0, critical=new_write, damaged=0, err_margin=err_margin,
+            total=total_written)
+        else:
+            self.log.info("Checking for %s less than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=0, critical=0, damaged=new_write, err_margin=err_margin,
```
Comment (kachhwahadivya) [inline libs/csm/rest/csm_rest_capacity.py:415]: new_write not required.

File: libs/csm/rest/csm_rest_capacity.py [hunk 2]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
...
+            total=total_written)
+        elif failure_cnt < kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=new_write, critical=0, damaged=0, err_margin=err_margin,
+            total=total_written)
+        elif failure_cnt == kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=0, critical=new_write, damaged=0, err_margin=err_margin,
```
Comment (kachhwahadivya) [inline libs/csm/rest/csm_rest_capacity.py:410]: new_write not required

File: libs/csm/rest/csm_rest_capacity.py [hunk 3]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
...
+        """
+        if failure_cnt == 0:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=total_written,
+            degraded=0, critical=0, damaged=0, err_margin=err_margin,
+            total=total_written)
+        elif failure_cnt < kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=new_write, critical=0, damaged=0, err_margin=err_margin,
```
Comment (kachhwahadivya) [inline libs/csm/rest/csm_rest_capacity.py:405]: new_write not required

File: libs/csm/rest/csm_rest_capacity.py [hunk 4]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
...
+        """
+        Verify degraded, critical, damaged bytecount in the resp.
+        """
+        if failure_cnt == 0:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=total_written,
+            degraded=0, critical=0, damaged=0, err_margin=err_margin,
+            total=total_written)
+        elif failure_cnt < kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
```
Comment (kachhwahadivya) [inline libs/csm/rest/csm_rest_capacity.py:403]: Can u correct the log?

File: tests/csm/rest/test_capacity_usage.py [hunk 5]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+                  obj, self.bucket, self.akey, self.skey, object_size=self.aligned_size)
+                assert resp, "Put object Failed"
+                total_written = total_written+self.aligned_size
+                result = self.csm_obj.verify_bytecount_fixed_placement(resp,failure_cnt, self.kvalue,
+                  test_cfg["err_margin"], total_written)
+                assert result[0], result[1]
+        self.log.info("[END] Failure loop")
+
+        self.log.info("[START] Recovery loop")
+        failure_cnt = len(self.failed_pod)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1234]: Add randomness

**Review by kachhwahadivya — COMMENTED** [review]
Can u revisit all the tests cases based on review comments.

File: libs/csm/rest/csm_rest_capacity.py [hunk 6]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
                 else:
                     self.log.error("Check for %s object in %s bucket incorrect.", obj, bucket)
         return checksum_match
+
+    def verify_bytecount_fixed_placement(self, resp, failure_cnt, kvalue, err_margin,
+                             total_written,new_write=0):
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply libs/csm/rest/csm_rest_capacity.py]: done

File: libs/csm/rest/csm_rest_capacity.py (hunk 1 above)
Comment (aditipande96) [reply libs/csm/rest/csm_rest_capacity.py]: done

File: libs/csm/rest/csm_rest_capacity.py (hunk 3 above)
Comment (aditipande96) [reply libs/csm/rest/csm_rest_capacity.py]: done

File: libs/csm/rest/csm_rest_capacity.py (hunk 4 above)
Comment (aditipande96) [reply libs/csm/rest/csm_rest_capacity.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 7]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+            self.failed_pod.append(deploy_name)
+
+            self.log.info("[Start] Check cluster status")
+            resp = self.ha_obj.check_cluster_status(self.master)
+            assert_utils.assert_false(resp[0], resp)
+            self.log.info("[End] Cluster is in degraded state")
+
+            self.log.info("[Start] Sleep %s", self.update_seconds)
+            time.sleep(self.update_seconds)
+            self.log.info("[Start] Sleep %s", self.update_seconds)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: libs/csm/rest/csm_rest_capacity.py [hunk 8]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
...
+            result = self.verify_degraded_capacity(resp, healthy=total_written,
+            degraded=0, critical=0, damaged=0, err_margin=err_margin,
+            total=total_written)
+        elif failure_cnt < kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=new_write, critical=0, damaged=0, err_margin=err_margin,
+            total=total_written)
+        elif failure_cnt == kvalue:
+            self.log.info("Checking for %s greater than to K value", failure_cnt)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply libs/csm/rest/csm_rest_capacity.py]: done

File: libs/csm/rest/csm_rest_capacity.py [hunk 9]
```diff
@@ -388,3 +388,30 @@ def verify_checksum(self, cap_df):
...
+            result = self.verify_degraded_capacity(resp, healthy=0,
+            degraded=0, critical=new_write, damaged=0, err_margin=err_margin,
+            total=total_written)
+        else:
+            self.log.info("Checking for %s less than to K value", failure_cnt)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply libs/csm/rest/csm_rest_capacity.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 10]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+    # pylint: disable-msg=too-many-statements
+    @pytest.mark.skip("Feature Not Ready")
+    @pytest.mark.lc
+    @pytest.mark.csmrest
+    @pytest.mark.cluster_user_ops
+    @pytest.mark.tags('TEST-44345')
+    def test_44345(self):
+        """
+        Test degraded capacity with Transient node failure in fixed protection scheme 
+		without write
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 11]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+    # pylint: disable-msg=too-many-statements
+    @pytest.mark.skip("Feature Not Ready")
+    @pytest.mark.lc
+    @pytest.mark.csmrest
+    @pytest.mark.cluster_user_ops
+    @pytest.mark.tags('TEST-44347')
+    def test_44347(self):
+        """
+        Test degraded capacity with Transient node failure in fixed protection scheme 
+		without write
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 12]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+            self.log.info("Deleting pod %s", deploy_name)
+            resp = self.master.create_pod_replicas(num_replica=0, deploy=deploy_name)
+            assert_utils.assert_false(resp[0], f"Failed to delete pod {deploy_name}")
+            self.log.info("[End] Successfully deleted pod %s", deploy_name)
+
+            self.failed_pod.append(deploy_name)
+
+            self.log.info("[Start] Check cluster status")
+            resp = self.ha_obj.check_cluster_status(self.master)
+            assert_utils.assert_false(resp[0], resp)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 13]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+            self.log.info("[Start] Sleep %s", self.update_seconds)
+            time.sleep(self.update_seconds)
+            self.log.info("[Start] Sleep %s", self.update_seconds)
+
+            resp = self.csm_obj.get_degraded_all(self.hlth_master)
+            index = deploy_name + "online"
+            new_row = pandas.Series(data=resp, name=index)
+            cap_df = cap_df.append(new_row, ignore_index=False)
+
+            result = self.csm_obj.verify_bytecount_fixed_placement(resp,failure_cnt, self.kvalue,
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 14]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
             break
         self.log.info("[END] completed")
         self.deploy = True
+
+    # pylint: disable-msg=too-many-statements
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py:1268]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 15]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+    def test_44342(self):
+        """
+        Test degraded capacity with Transient node failure in fixed protection scheme 
+		without write
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        test_cfg = self.csm_conf["test_44342"]
+        cap_df = pandas.DataFrame()
+        resp = self.csm_obj.get_degraded_all(self.hlth_master)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: tested with the existing code and it passed

File: tests/csm/rest/test_capacity_usage.py [hunk 16]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+            new_row = pandas.Series(data=resp, name=index)
+            cap_df = cap_df.append(new_row, ignore_index=False)
+
+            result = self.csm_obj.verify_bytecount_fixed_placement(resp,failure_cnt, self.kvalue,
+                test_cfg["err_margin"], total_written)
+            assert result[0], result[1]
+        self.log.info("[END] Failure loop")
+
+        self.log.info("[START] Recovery loop")
+        failure_cnt = len(self.failed_pod)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py:1323]: done

File: tests/csm/rest/test_capacity_usage.py (hunk 5 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py:1425]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 17]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+            self.log.info("[Start] Sleep %s", self.update_seconds)
+
+            resp = self.csm_obj.get_degraded_all(self.hlth_master)
+            index = deploy_name + "online"
+            new_row = pandas.Series(data=resp, name=index)
+            cap_df = cap_df.append(new_row, ignore_index=False)
+
+            result = self.csm_obj.verify_bytecount_fixed_placement(resp,failure_cnt, self.kvalue,
+                test_cfg["err_margin"], total_written)
+            assert result[0], result[1] + f"for {failure_cnt} failures"
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 18]
```diff
@@ -1084,3 +1084,267 @@ def test_39923(self):
...
+    @pytest.mark.tags('TEST-44342')
+    def test_44342(self):
+        """
+        Test degraded capacity with Transient node failure in fixed protection scheme 
+		without write
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        test_cfg = self.csm_conf["test_44342"]
+        cap_df = pandas.DataFrame()
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py [hunk 19]
```diff
@@ -1084,3 +1085,437 @@ def test_39923(self):
...
+            cls.log.debug("Worker node : %s", node["hostname"])
+            cls.node_list.append(LogicalNode(hostname=node["hostname"],
+                                                 username=node["username"],
+                                                 password=node["password"]))
+            host = node["hostname"]
+            cls.host_list.append(host)
+        cls.log.info("Worker node List: %s", cls.host_list)
+        cls.num_worker = len(cls.host_list)
+        cls.log.info("Number of workers detected: %s", cls.num_worker)
+        cls.nd_obj = LogicalNode(hostname=CMN_CFG["nodes"][0]["hostname"],
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1125]: Not required

File: tests/csm/rest/test_capacity_usage.py [hunk 20]
```diff
@@ -1084,3 +1085,437 @@ def test_39923(self):
...
+        cls.s3_user = ""
+        cls.bucket = ""
+        cls.row_temp = "N{} failure"
+        cls.node_list = []
+        cls.host_list = []
+        cls.num_nodes = len(CMN_CFG["nodes"])
+        cls.io_bucket_name = f"iobkt1-copyobject-{time.perf_counter_ns()}"
+        cls.s3_obj = s3_test_lib.S3TestLib()
+        cls.ha_obj = HAK8s()
+        for node in CMN_CFG["nodes"]:
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1115]: Not required

File: tests/csm/rest/test_capacity_usage.py [hunk 21]
```diff
@@ -1084,3 +1085,437 @@ def test_39923(self):
...
+        self.akey = resp.json()["access_key"]
+        self.skey = resp.json()["secret_key"]
+        self.s3_user = resp.json()["account_name"]
+        self.bucket = "iam-user-bucket-" + str(int(time.time()))
+
+        self.log.info("Verify Create bucket: %s with access key: %s and secret key: %s",
+                      self.bucket, self.akey, self.skey)
+        assert s3_misc.create_bucket(self.bucket, self.akey, self.skey), "Failed to create bucket."
+
+        self.cap_df = pandas.DataFrame()
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1178]: Not required

File: tests/csm/rest/test_capacity_usage.py (hunk 20 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py:1115]: This is required for node_list variable which is further used in teardown

File: tests/csm/rest/test_capacity_usage.py (hunk 21 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: removed

File: tests/csm/rest/test_capacity_usage.py (hunk 19 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: removed

File: tests/csm/rest/test_capacity_usage.py [hunk 22]
```diff
@@ -1084,3 +1085,423 @@ def test_39923(self):
...
+                resp = self.csm_obj.get_degraded_all(self.csm_obj.hlth_master)
+
+                result = self.csm_obj.verify_bytecount_fixed_placement(resp, failure_cnt,
+                         self.kvalue, self.err_margin, total_written)
+                assert result[0], result[1]
+        self.log.info("[END] Failure loop")
+
+        self.log.info("[START] Recovery loop")
+        failure_cnt = len(self.failed_pod)
+        random.shuffle(self.failed_pod)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1486]: use the random object from test_lib

File: tests/csm/rest/test_capacity_usage.py [hunk 23]
```diff
@@ -1084,3 +1084,423 @@ def test_39923(self):
...
+            cls.log.debug("Worker node : %s", node["hostname"])
+            cls.node_list.append(LogicalNode(hostname=node["hostname"],
+                                                 username=node["username"],
+                                                 password=node["password"]))
+        cls.restore_pod = None
+        cls.restore_method = RESTORE_SCALE_REPLICAS
+        cls.deployment_name = []
+        cls.failed_pod = []
+        cls.deployment_backup = None
+        cls.fail_cnt = 0
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1122]: Unused variable. Can u please scan through the class and remove all unused variable

File: tests/csm/rest/test_capacity_usage.py [hunk 24]
```diff
@@ -1084,3 +1084,423 @@ def test_39923(self):
...
+        cls.bucket = ""
+        cls.node_list = []
+        cls.num_nodes = len(CMN_CFG["nodes"])
+        cls.io_bucket_name = f"iobkt1-copyobject-{time.perf_counter_ns()}"
+        cls.s3_obj = s3_test_lib.S3TestLib()
+        cls.ha_obj = HAK8s()
+        for node in CMN_CFG["nodes"]:
+            cls.log.debug("Worker node : %s", node["hostname"])
+            cls.node_list.append(LogicalNode(hostname=node["hostname"],
+                                                 username=node["username"],
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1115]: Can u run autopep for correcting all PEP8 mistakes?

File: tests/csm/rest/test_capacity_usage.py (hunk 20 above)
Comment (kachhwahadivya) [reply tests/csm/rest/test_capacity_usage.py:1112]: This is available in test lib

File: tests/csm/rest/test_capacity_usage.py [hunk 25]
```diff
@@ -1084,3 +1084,423 @@ def test_39923(self):
...
+        cls.user_pass = cls.csm_obj.config["csm_admin_user"]["password"]
+        cls.deploy = False
+        cls.akey = ""
+        cls.skey = ""
+        cls.s3_user = ""
+        cls.bucket = ""
+        cls.node_list = []
+        cls.num_nodes = len(CMN_CFG["nodes"])
+        cls.io_bucket_name = f"iobkt1-copyobject-{time.perf_counter_ns()}"
+        cls.s3_obj = s3_test_lib.S3TestLib()
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1110]: Not required

File: tests/csm/rest/test_capacity_usage.py [hunk 26]
```diff
@@ -1084,3 +1084,423 @@ def test_39923(self):
...
+        cls.username = cls.csm_obj.config["csm_admin_user"]["username"]
+        cls.user_pass = cls.csm_obj.config["csm_admin_user"]["password"]
+        cls.deploy = False
+        cls.akey = ""
+        cls.skey = ""
+        cls.s3_user = ""
+        cls.bucket = ""
+        cls.node_list = []
+        cls.num_nodes = len(CMN_CFG["nodes"])
+        cls.io_bucket_name = f"iobkt1-copyobject-{time.perf_counter_ns()}"
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1109]: Unused

File: tests/csm/rest/test_capacity_usage.py [hunk 27]
```diff
@@ -1084,3 +1084,423 @@ def test_39923(self):
...
+        if not self.deploy:
+            bundle_dir = os.path.join(LOG_DIR, "latest", "support_bundle")
+            self.log.info("Support bundle dir : %s", bundle_dir)
+            resp = sb.collect_support_bundle_k8s(local_dir_path=bundle_dir,
+                                                scripts_path=K8S_SCRIPTS_PATH)
+
+        self.deploy = True
+        if self.deploy:
+            self.log.info("Cleanup: Destroying the cluster ")
+            resp = self.deploy_lc_obj.destroy_setup(self.csm_obj.master, self.node_list,
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_usage.py:1221]: self.node_list is same as worker_list in test_lib

File: tests/csm/rest/test_capacity_usage.py [hunk 28]
```diff
@@ -1084,3 +1084,423 @@ def test_39923(self):
...
+                             K8S_SCRIPTS_PATH)
+            assert_utils.assert_true(resp[0], resp[1])
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py (hunk 27 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: replaced

File: tests/csm/rest/test_capacity_usage.py (hunk 26 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py (hunk 25 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py (hunk 24 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: committed autopep8 changes for entire file

File: tests/csm/rest/test_capacity_usage.py (hunk 23 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py (hunk 20 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

File: tests/csm/rest/test_capacity_usage.py (hunk 27 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_usage.py]: done

**Review by kachhwahadivya — APPROVED** [review]

**Comment by aditipande96** [comment]
========= 2339 tests collected in 10.50s ===========

**Review by RajeshPratapSInghChouhan — APPROVED** [review]

**Review by ravindrachoudhari — APPROVED** [review]

**Pull request merged**